File: C++/1812. 【认证试题】矩阵列的最大值中的最小值.cpp
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2019-2020. All rights reserved.
 * Description: 考生实现代码
 * Note: 缺省代码仅供参考，可自行决定使用、修改或删除
*/
#include <iostream>
#include <sstream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    // 待实现函数，在此函数中填入答题代码;
    int GetTheMinWithColMaxs(const vector<vector<int>> &matrix) 
    {
        int res = (1<<31) - 1;
        for(int j=0;j<matrix[0].size();++j) {
            int temp = -10001;
            for(int i=0;i<matrix.size();++i) {
                temp = max(temp, matrix[i][j]);
            }
            res =  min(res, temp);
        }
        return res;
    }
};

int main()
{
    int row, col;
    cin >> row >> col;
    cin.ignore();
  
    vector<vector<int>> matrix;
    for (int loop = 0; loop < row; loop++) {
        string line;
        getline(cin, line);
      
        istringstream iss(line);
        vector<int> rows;
        while (!iss.eof()) {
            int value;
            iss >> value;
            rows.push_back(value);
        }
        matrix.push_back(rows);
    }

    Solution solu;
    int result = solu.GetTheMinWithColMaxs(matrix);
    cout << result << endl;  
    
    return 0;
}

File: Python/1812. 【认证试题】矩阵列的最大值中的最小值.py
"""
Copyright (c) Huawei Technologies Co., Ltd. 2020-2020. All rights reserved.
Description: 请实现代码
Note: 缺省代码仅供参考，可自行决定使用、修改或删除
"""


class Solution:
    def get_the_min_with_col_maxs(self, matrix):
        res = []
        
        for j in range(len(matrix[0])):
            temp = -10001
            for i in range(len(matrix)):
                temp = max(temp, matrix[i][j])
            res.append(temp)
        return min(res)

if __name__ == "__main__":
    n, m = map(int, input().strip().split())
    matrix = [list(map(int, input().strip().split())) for _ in range(n)]
    function = Solution()
    results = function.get_the_min_with_col_maxs(matrix)
    print(results)

